downstairs/src/region.rs

```diff
@@ -1165,6 +1187,13 @@ impl Extent {
             }
         }
 
+        // We do all of our metadb updates in a single transaction to minimize
+        // syncs.  (Note that the "unchecked" in the signature merely denotes
+        // that we are taking responsibility for assuring that we are not
+        // in a nested transaction, accepting that it will fail at run-time
+        // if we are.)
+        let tx = inner.metadb.unchecked_transaction()?;
+
         inner.set_dirty()?;
```

Comment:
Could we hoist the dtrace probe point just above this?

I'm also wondering about the `set_dirty()` here, and if it should come at the top as well.  I'm not sure all the failure cases of failing on the `unchecked_transaction()`, like, in the failure case, will it for sure not have dirtied something that we will need to repair?  It seems like it's okay, but missing a dirty bit setting can result in a bad repair and data corruption that could be very very hard to find, so setting that dirty bit correctly is critical.

Reply:
On the DTrace probe: which one do you want to hoist?  (Note that this change isn't changing the `set_dirty` with respect to the existing `extent-write-sqlite-insert-start` probe.)

Reply:
>

Reply:
> So, per the block comment immediately above it, `unchecked_transaction` is a bit poorly named. The reason it will fail is due to programmer error (and I honestly wish that this wouldn't fail but would instead `panic` on failure); the failure case isn't something that we would expect.

You could do this yourself (i.e., check for the "nested transaction" error and panic), but it's super awkward. Opening a transaction executes a SQL statement so can (at the API level) fail for any reason any SQL execution could fail, and the failure on nested transactions specifically does not have its own error code:

```
SqliteFailure(Error { code: Unknown, extended_code: 1 }, Some("cannot start a transaction within a transaction"))
```

`extended_code: 1` is [`SQLITE_ERROR`](https://www.sqlite.org/rescode.html#error):

> The SQLITE_ERROR result code is a generic error code that is used when no other more specific error code is available.

which means you'd have to check the error string (which itself is coming from sqlite) :nauseated_face:

Reply:
Ah, live discussion had informed me, and I think we are good to go like this was.